contracts/eosio.system/include/eosio.system/eosio.system.hpp

```diff
@@ -337,6 +339,31 @@ namespace eosiosystem {
 
    typedef eosio::multi_index< "rexpool"_n, rex_pool > rex_pool_table;
 
+   // `rex_return_pool` structure underlying the rex return pool table. A rex return pool table entry is defined by:
+   // - `version` defaulted to zero,
+   // - `last_update_time` the last time returns from renting, ram fees, and name bids were added to the rex pool,
+   // - `current_rate_of_increase` current amount to be added to the rex pool at a rate of per 30 days,
+   // - `cummulative_proceeds` bookkeeping variable that tracks fees added to rex pool up to current time,
+   // - `proceeds` bookkeeping variable that tracks fees added to rex return pool up to current time,
+   // - `return_buckets` 12-hour buckets containing amounts to be added to the rex pool and the times they become effective 
+   struct [[eosio::table,eosio::contract("eosio.system")]] rex_return_pool {
+      uint8_t                       version = 0;
+      
+      time_point                    last_update_time;
+      int64_t                       current_rate_of_increase = 0;
+      int64_t                       cummulative_proceeds = 0;
+      int64_t                       proceeds = 0;
+      std::map<time_point, int64_t> return_buckets;
```

Comment:
The value will need to have two fields rather than a single `int64_t`. One tracks the `proceeds` (not to be confused by the global `proceeds` above) accumulated from renting and fees during the time the bucket is active. The other is the distribution rate per `dist_interval` time interval that is calculated after the `proceeds` field in the same bucket becomes immutable.

Reply:
Or better yet, the `proceeds` can just be kept in the singleton object and the value of the map only needs the rate. In fact, if you withhold constructing the bucket until after the end time of the bucket time interval, the buckets can be considered immutable. Also I think the bucket map should be in its own object because it can be a significant amount of data that is updated infrequently (once every 12 hours) while the rest of the fields in the singleton object could be updated many times per block.